File: examples/Training/python/Create_Image.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (C) 2014 University of Dundee & Open Microscopy Environment.
#                    All Rights Reserved.
# Use is subject to license terms supplied in LICENSE.txt
#

"""
FOR TRAINING PURPOSES ONLY!
"""

from omero.gateway import BlitzGateway
from Parse_OMERO_Properties import USERNAME, PASSWORD, HOST, PORT
from Parse_OMERO_Properties import imageId

"""
start-code
"""

# Create a connection
# ===================
conn = BlitzGateway(USERNAME, PASSWORD, host=HOST, port=PORT)
conn.connect()


# Create an image from scratch
# ============================
# This example demonstrates the usage of the convenience method
# createImageFromNumpySeq() Here we create a multi-dimensional image from a
# hard-coded array of data.
from numpy import array, int8
import omero
size_x, size_y, size_z, size_c, size_t = 5, 4, 1, 2, 1
plane1 = array(
    [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [0, 1, 2, 3, 4], [5, 6, 7, 8, 9]],
    dtype=int8)
plane2 = array(
    [[5, 6, 7, 8, 9], [0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [0, 1, 2, 3, 4]],
    dtype=int8)
planes = [plane1, plane2]


def plane_gen():
    """generator will yield planes"""
    for p in planes:
        yield p


desc = "Image created from a hard-coded arrays"
i = conn.createImageFromNumpySeq(
    plane_gen(), "numpy image", size_z, size_c, size_t, description=desc,
    dataset=None)

print 'Created new Image:%s Name:"%s"' % (i.getId(), i.getName())


# Set the pixel size using units
# ==============================
# Lengths are specified by value and a unit enumeration
# Here we set the pixel size X and Y to be 9.8 Angstroms

from omero.model.enums import UnitsLength
# Re-load the image to avoid update conflicts
i = conn.getObject("Image", i.getId())
u = omero.model.LengthI(9.8, UnitsLength.ANGSTROM)
p = i.getPrimaryPixels()._obj
p.setPhysicalSizeX(u)
p.setPhysicalSizeY(u)
conn.getUpdateService().saveObject(p)


# Create an Image from an existing image
# ======================================
# We are going to create a new image by passing the method a 'generator' of 2D
# planes This will come from an existing image, by taking the average of 2
# channels.
zct_list = []
image = conn.getObject('Image', imageId)
size_z, size_c, size_t = image.getSizeZ(), image.getSizeC(), image.getSizeT()
dataset = image.getParent()
pixels = image.getPrimaryPixels()
new_size_c = 1


def plane_gen():
    """
    set up a generator of 2D numpy arrays.

    The createImage method below expects planes in the order specified here
    (for z.. for c.. for t..)
    """
    for z in range(size_z):              # all Z sections
        # Illustrative purposes only, since we only have 1 channel
        for c in range(new_size_c):
            for t in range(size_t):      # all time-points
                channel0 = pixels.getPlane(z, 0, t)
                channel1 = pixels.getPlane(z, 1, t)
                # Here we can manipulate the data in many different ways. As
                # an example we are doing "average"
                # average of 2 channels
                new_plane = (channel0 + channel1) / 2
                print "newPlane for z,t:", z, t, new_plane.dtype, \
                    new_plane.min(), new_plane.max()
                yield new_plane
desc = ("Image created from Image ID: %s by averaging Channel 1 and Channel 2"
        % imageId)
i = conn.createImageFromNumpySeq(
    plane_gen(), "new image", size_z, new_size_c, size_t, description=desc,
    dataset=dataset)


# Close connection
# ================
# When you are done, close the session to free up server resources.
conn.close()

File: examples/Training/matlab/CreateImage.m
% Copyright (C) 2011-2014 University of Dundee & Open Microscopy Environment.
% All rights reserved.
%
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License along
% with this program; if not, write to the Free Software Foundation, Inc.,
% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

% Crate a new synthetic image and upload it to the server

try
    % Initialize a client and a session using the ice.config file
    % See ConnectToOMERO for alternative ways to initialize a session
    [client, session] = loadOmero();
    p = parseOmeroProperties(client);
    eventContext = session.getAdminService().getEventContext();
    fprintf(1, 'Created connection to %s\n', p.hostname);
    msg = 'Created session for user %s (id: %g) using group %s (id: %g)\n';
    fprintf(1, msg, char(eventContext.userName), eventContext.userId,...
        char(eventContext.groupName), eventContext.groupId);
    
    % Information to edit
    datasetId = p.datasetid;
    
    % Read the dimensions
    sizeX = 200;
    sizeY = 100;
    sizeZ = 1; % The number of z-sections.
    sizeT = 5; % The number of timepoints.
    sizeC = 2; % The number of channels.
    type = 'uint16';
    
    % Retrieve pixel type
    pixelsService = session.getPixelsService();
    pixelTypes = toMatlabList(pixelsService.getAllEnumerations('omero.model.PixelsType'));
    pixelTypeValues = arrayfun(@(x) char(x.getValue().getValue()),...
        pixelTypes, 'Unif', false);
    pixelType = pixelTypes(strcmp(pixelTypeValues, type));
    
    % Create a new image
    disp('Uploading new image onto the server');
    description = sprintf('Dimensions: %g x %g x %g x %g x %g',...
        sizeX, sizeY, sizeZ, sizeC, sizeT);
    name = 'New image';
    idNew = pixelsService.createImage(sizeX, sizeY, sizeZ, sizeT,...
        toJavaList(0:sizeC-1, 'java.lang.Integer'), pixelType, name, description);
    
    %load the image.
    disp('Checking the created image');
    imageNew = getImages(session, idNew.getValue());
    assert(~isempty(imageNew), 'OMERO:CreateImage', 'Image Id not valid');
    
    % Set channel properties
    disp('Adding metadata to the channels')
    channels = loadChannels(session, imageNew);
    for i = 1: numel(channels)
        channelName = ['Channel ' num2str(i)'];
        emissionWave = omero.model.LengthI;
        emissionWave.setValue(550);
        emissionWave.setUnit(omero.model.enums.UnitsLength.NANOMETER);
        channels(i).getLogicalChannel().setName(rstring(channelName));
        channels(i).getLogicalChannel().setEmissionWave(emissionWave);
    end
    session.getUpdateService().saveArray(toJavaList(channels));
    
    % load the dataset
    fprintf(1, 'Reading dataset: %g\n', datasetId);
    dataset = getDatasets(session, datasetId, false);
    assert(~isempty(dataset), 'OMERO:CreateImage', 'Dataset Id not valid');
    
    % Link the new image to the dataset
    fprintf(1, 'Linking image %g to dataset %g\n', idNew.getValue(), datasetId);
    link = omero.model.DatasetImageLinkI;
    link.setChild(omero.model.ImageI(idNew, false));
    link.setParent(omero.model.DatasetI(dataset.getId().getValue(), false));
    session.getUpdateService().saveAndReturnObject(link);
    
    % Copy the data.
    fprintf(1, 'Copying data to image %g\n', idNew.getValue());
    pixels = imageNew.getPrimaryPixels();
    store = session.createRawPixelsStore();
    store.setPixelsId(pixels.getId().getValue(), false);
    
    % Create template for upload
    range =  cast(1: intmax(type) / sizeX  : intmax(type), type);
    template = repmat(range', 1, sizeY);
    byteArray = toByteArray(template, pixels);
    
    % Upload template for every plane in the image
    for z = 1 : sizeZ,
        for c = 1:sizeC
            for t = 1: sizeT,
                index = sub2ind([sizeZ sizeC sizeT], z, c, t);
                store.setPlane(byteArray, z - 1, c - 1, t - 1);
            end
        end
    end
    store.save(); %save the data
    store.close(); %close
    
catch err
    client.closeSession();
    throw(err);
end

% Close the session
client.closeSession();
